src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +429,411 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+  /-- Only perform a round of shift-subtract if we have dividend bits. -/
+  hwn_lt : 0 < wn
+
+/--
+In the shift subtract input, the dividend is at least one bit long (`wn > 0`), so
+the remainder has bits to be computed (`wr < w`).
+-/
+def DivModState.wr_lt_w {qr : DivModState w} (h : qr.LawfulShiftSubtract wr wn n d) : wr < w := by
+  have hwrn := h.hwrn
+  have hwn_lt := h.hwn_lt
+  omega
+
+/-- If we have extra bits to spare in `n`,
+then the div rem input can be converted into a shift subtract input
+to run a round of the shift subtracter. -/
+def DivModState.Lawful.toLawfulShiftSubtract {qr : DivModState w}
+    (h : qr.Lawful w wr (wn + 1) n d) : qr.LawfulShiftSubtract wr (wn + 1) n d where
+  hwrn := by have := h.hwrn; omega
+  hdPos := h.hdPos
+  hrLtDivisor := h.hrLtDivisor
+  hrWidth := h.hrWidth
+  hqWidth := h.hqWidth
+  hdiv := h.hdiv
+  hwn_lt := by omega
+
+/-! ### shiftConcat -/
+
+/--
+If `n : Bitvec w` has only the low `k < w` bits set,
+then `(n <<< 1 | b)` does not overflow.
+-/
+private theorem mul_two_add_lt_two_pow_of_lt_two_pow_of_lt_two {n b k w : Nat}
+    (hn : n < 2 ^ k) (hb : b < 2) (hk : k < w) :
+    n * 2 + b < 2 ^ w := by
+  have : 2^(k + 1) ≤ 2 ^w := Nat.pow_le_pow_of_le_right (by decide) (by assumption)
+  omega
+
+/-- `x.shiftConcat b` does not overflow if `x < 2^k` for `k < w`, and so
+`x.shiftConcat b |>.toNat = x.toNat * 2 + b.toNat`. -/
+theorem toNat_shiftConcat_eq_of_lt_of_lt_two_pow {x : BitVec w} {b : Bool} {k : Nat}
+  (hk : k < w) (hx : x.toNat < 2 ^ k) :
+    (x.shiftConcat b).toNat = x.toNat * 2 + b.toNat := by
+  simp only [shiftConcat]
+  have : b.toNat = if b then 1 else 0 := by rcases b <;> rfl
+  rw [this]
+  have hk' : 2^k < 2^w := by
+    apply Nat.pow_lt_pow_of_lt (by decide) (by omega)
+  rcases w with rfl | w
+  · omega -- contradiction, k < w
+  · rw [← BitVec.add_eq_or_of_and_eq_zero]
+    · simp only [toNat_add, toNat_shiftLeft, toNat_truncate, toNat_ofBool, toNat, add_mod_mod,
+      mod_add_mod]
+      rw [Nat.shiftLeft_eq]
+      simp only [show (2 ^ 1 = 2) by decide]
+      rw [Nat.mod_eq_of_lt]
+      · rcases b with rfl | rfl <;> simp
+      · apply mul_two_add_lt_two_pow_of_lt_two_pow_of_lt_two
+        · exact hx
+        · rcases b <;> decide
+        · assumption
+    · ext i
+      simp
+      omega
+
+theorem toNat_shiftConcat_lt_of_lt_of_lt_two_pow {x : BitVec w} {b : Bool} {k : Nat}
+    (hk : k < w) (hx : x.toNat < 2 ^ k) :
+    (x.shiftConcat b).toNat < 2 ^ (k + 1) := by
+  rw [toNat_shiftConcat_eq_of_lt_of_lt_two_pow hk hx]
+  apply mul_two_add_lt_two_pow_of_lt_two_pow_of_lt_two hx
+  · cases b <;> decide
+  · omega
+
+@[bv_toNat]
+theorem toNat_shiftConcat {x : BitVec w} {b : Bool} : (x.shiftConcat b).toNat =
+    (x.toNat <<< 1 + b.toNat) % 2 ^ w  := by
+  simp only [shiftConcat]
+  rw [← add_eq_or_of_and_eq_zero] -- Due to `add_eq_or_of_and_eq_zero`, this must live in `Bitblast`.
+  · simp
+  · ext i
+    simp
+    omega
```

Comment:
DOes this belong in `BitVec/Lemmas.lean`? Also, this should have @[simp] as well.

Reply:
Sadly not, as the comment says: `Due to `add_eq_or_of_and_eq_zero`, this must live in `Bitblast`. `